src/Features/CSharp/Portable/ConvertAutoPropertyToFullProperty/CSharpConvertAutoPropertyToFullPropertyCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,226 @@
+            }
+
+            return containingProperty;
+        }
+
+        internal override async Task<string> GetFieldNameAsync(Document document, IPropertySymbol propertySymbol, CancellationToken cancellationToken)
+        {
+            var rules = await GetNamingRulesAsync(document, cancellationToken).ConfigureAwait(false);
+            return GenerateFieldName(propertySymbol, rules);
+        }
+
+        /// <summary>
+        /// Get the user-specified naming rules, then add standard default naming rules 
+        /// for both static and non-static fields.  The standard naming rules are added at the end 
+        /// so they will only be used if the user hasn't specified a preference.
+        /// </summary>
+        private static async Task<ImmutableArray<NamingRule>> GetNamingRulesAsync(
+            Document document,
+            CancellationToken cancellationToken)
+        {
+            const string defaultStaticFieldPrefix = "s_";
+            const string defaultFieldPrefix = "_";
+
+            var optionSet = await document.GetOptionsAsync(cancellationToken).ConfigureAwait(false);
+            var namingPreferencesOption = optionSet.GetOption(SimplificationOptions.NamingPreferences);
+            var rules = namingPreferencesOption.CreateRules().NamingRules
+                .AddRange(CreateNewRule(ImmutableArray.Create(new ModifierKind(ModifierKindEnum.IsStatic)), defaultStaticFieldPrefix))
+                .AddRange(CreateNewRule(ImmutableArray.Create<ModifierKind>(), defaultFieldPrefix));
+            return rules;
+        }
+
+        private static ImmutableArray<NamingRule> CreateNewRule(
+            ImmutableArray<ModifierKind> modifiers,
+            string prefix)
+        {
+            return ImmutableArray.Create(
+                new NamingRule(
+                    new SymbolSpecification(
+                        Guid.NewGuid(),
+                        "Field",
+                        ImmutableArray.Create(new SymbolSpecification.SymbolKindOrTypeKind(SymbolKind.Field)),
+                        modifiers: modifiers),
+                    new NamingStyles.NamingStyle(
+                        Guid.NewGuid(),
+                        prefix: prefix,
+                        capitalizationScheme: Capitalization.CamelCase),
+                    DiagnosticSeverity.Hidden));
+        }
+
+        private string GenerateFieldName(IPropertySymbol property, ImmutableArray<NamingRule> rules)
+        {
+            var propertyName = property.Name;
+            var fieldName = "";
+            foreach (var rule in rules)
+            {
+                if (rule.SymbolSpecification.AppliesTo(
+                    new SymbolKindOrTypeKind(SymbolKind.Field),
+                    property.IsStatic ? DeclarationModifiers.Static : DeclarationModifiers.None,
+                    Accessibility.Private))
+                {
+                    fieldName = rule.NamingStyle.MakeCompliant(propertyName).First();
```

Comment:
nit: @dpoeschl it seems unfortunate to have to allocate all these iterators because MakeCompliant sometimes wants to return two values.  Consider having MakeCompliant return a tuple in the future.